### DISCUSSION TITLE
How to configure custom analytics - 51la

### DISCUSSION THREAD

**Discussion by w1ndys** [squidfunk/mkdocs-material#6888] [Q&A] [answered]
I have arranged the js footer file, and it seems to be error-free, but on the 51la platform, it can't pass the code detection and displays "No code detected".

My project address
https://github.com/W1ndys/xkzb.qfnu.w1ndys.top

https://v6.51.la/
![Screenshot_2024-03-09-02-21-12-941_com android chrome](https://github.com/squidfunk/mkdocs-material/assets/109416673/41f522b3-ecc0-4d1a-89aa-b9ba097a0414)

**Comment by w1ndys** [comment]
I suspect it's the location,Another question is how do I add js to the <head/> tag?

**Reply by kamilkrzyskow** [reply]
> Another question is how do I add js to the tag?

You can add custom `<script>` tags in your `main.html` template override, in the `libs` or `extrahead` blocks:
https://squidfunk.github.io/mkdocs-material/customization/#overriding-blocks

But there is also the `custom.html` analytics partial:
https://squidfunk.github.io/mkdocs-material/setup/setting-up-site-analytics/#custom-site-analytics

**Comment by kamilkrzyskow** [comment]
Looking at the source `view-source:https://xkzb.qfnu.w1ndys.top/` you can see that all files are properly linked 
```html
<script src="assets/javascripts/bundle.c8d2eff1.min.js"></script>
<script src="javascripts/extra.js"></script>
<script src="javascripts/51la.js"></script>
```
However, the issue is that CORS is blocking the connection: 
![image](https://github.com/squidfunk/mkdocs-material/assets/34622465/c63823bf-8541-492b-a7ad-935a3b34549d)
Seen in the browser Dev Tools console.
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSDidNotSucceed

I can't read Chinese, so it would be good if you've provided what 51.la is, but ChatGPT says it's some sort of tracking analytics thing. 
I guess you have to somehow whitelist your domain `https://xkzb.qfnu.w1ndys.top/` on some sort of management back panel in your 51.la account.

This is definitely outside the scope of this discussions board, so best look for help in 51.la forums :v:

**Reply by w1ndys** [reply]
Thank you. I saw in the background that it seems to be running normally. I don't know what the reason was before. Maybe there was a delay in code detection.

In addition, I have another question. In the official documentation of 51la, it is recommended to insert script into the `<head>` tag. In material, how to insert external js into the `<head>` tag?

**Reply by kamilkrzyskow** [reply, marked as answer]
Instead of using the `extra_javascript` in mkdocs.yml, use `main.html` override with `extrahead` block, as said in the comment above:
- https://github.com/squidfunk/mkdocs-material/discussions/6888#discussioncomment-8726317

As for the syntax look here for reference:
- https://github.com/squidfunk/mkdocs-material/discussions/6881

**Reply by w1ndys** [reply]
ohhhhh, ok  thanks!!!!

**Reply by w1ndys** [reply]
Thank you very much for your help, I have implemented this function, I have added it to the `<head>` tag

But I still have some questions about the syntax of `custom.html`.
What do `provide` and `property` stand for respectively? `property` may represent the authentication ID of site statistics? Can I create html at will inside the `overrides` directory and import it into `mkdocs.yml`

For example:
I created an `overrides/W1ndys1/W1ndys2.html`
Then configure it in `mkdocs.yml` as follows:

```yml
extra:
   W1ndys1:
     provider: W1ndys2
     .....
```
Is this possible?
> I tested it and found that it seemed to have no effect. Does this prove that my writing method is wrong?

I saw some explanations of variables in the [document](https://squidfunk.github.io/mkdocs-material/customization/#overriding-blocks), but I am not sure whether these variables are only used to write main.html or whether all html in overrides need to follow these variable syntax.

![image](https://github.com/squidfunk/mkdocs-material/assets/109416673/1aae2296-8c7c-4628-967f-1c0824873df4)

**Reply by w1ndys** [reply]
I seem to understand, does it only support the following content?
```
.
├─ .icons/                             # Bundled icon sets
├─ assets/
│  ├─ images/                          # Images and icons
│  ├─ javascripts/                     # JavaScript files
│  └─ stylesheets/                     # Style sheets
├─ partials/
│  ├─ integrations/                    # Third-party integrations
│  │  ├─ analytics/                    # Analytics integrations
│  │  └─ analytics.html                # Analytics setup
│  ├─ languages/                       # Translation languages
│  ├─ actions.html                     # Actions
│  ├─ alternate.html                   # Site language selector
│  ├─ comments.html                    # Comment system (empty by default)
│  ├─ consent.html                     # Consent
│  ├─ content.html                     # Page content
│  ├─ copyright.html                   # Copyright and theme information
│  ├─ feedback.html                    # Was this page helpful?
│  ├─ footer.html                      # Footer bar
│  ├─ header.html                      # Header bar
│  ├─ icons.html                       # Custom icons
│  ├─ language.html                    # Translation setup
│  ├─ logo.html                        # Logo in header and sidebar
│  ├─ nav.html                         # Main navigation
│  ├─ nav-item.html                    # Main navigation item
│  ├─ pagination.html                  # Pagination (used for blog)
│  ├─ palette.html                     # Color palette toggle
│  ├─ post.html                        # Blog post excerpt
│  ├─ progress.html                    # Progress indicator
│  ├─ search.html                      # Search interface
│  ├─ social.html                      # Social links
│  ├─ source.html                      # Repository information
│  ├─ source-file.html                 # Source file information
│  ├─ tabs.html                        # Tabs navigation
│  ├─ tabs-item.html                   # Tabs navigation item
│  ├─ tags.html                        # Tags
│  ├─ toc.html                         # Table of contents
│  ├─ toc-item.html                    # Table of contents item
│  └─ top.html                         # Back-to-top button
├─ 404.html                            # 404 error page
├─ base.html                           # Base template
├─ blog.html                           # Blog index page
├─ blog-archive.html                   # Blog archive index page
├─ blog-category.html                  # Blog category index page
├─ blog-post.html                      # Blog post page
└─ main.html                           # Default page
```
And the above picture describes only the writing methods supported in `main.html`

**Reply by kamilkrzyskow** [reply]
> but I am not sure whether these variables are only used to write main.html or whether all html in overrides need to follow these variable syntax.

If by "variables" you mean "blocks" then they are limited to the `main.html` file. All of the blocks are defined in the `base.html`
https://github.com/squidfunk/mkdocs-material/blob/fdeafbb875fe7cf340da429c90e8eaa965e9663e/src/templates/base.html#L172-L173
and `main.html` extends the `base.html` allowing to override the blocks with additional content or fully replace them.
https://github.com/squidfunk/mkdocs-material/blob/fdeafbb875fe7cf340da429c90e8eaa965e9663e/src/templates/main.html#L23

Basically `main.html` is used to build your docs, and loads the `base.html` in an action chain.

---

> But I still have some questions about the syntax of custom.html.
What do provide and property stand for respectively? property may represent the authentication ID of site statistics? Can I create html at will inside the overrides directory and import it into mkdocs.yml

The `base.html` template loads the built-in `analytics.html` partial:
https://github.com/squidfunk/mkdocs-material/blob/fdeafbb875fe7cf340da429c90e8eaa965e9663e/src/templates/base.html#L156-L159
and later the `analytics.html` partial loads other files based on the values provided in `mkdocs.yml`
https://github.com/squidfunk/mkdocs-material/blob/fdeafbb875fe7cf340da429c90e8eaa965e9663e/src/templates/partials/integrations/analytics.html#L23-L30
So after setting:
```yml
extra:
  analytics:
    provider: google
    property: G-XXXXXXXXXX
```
the `config.extra.analytics.provider` will hold the `google` value and later it will be used to include the "partials/integrations/analytics/google.html" file. Later the `google.html` partial uses the `property` to include it within the analytics script.
https://github.com/squidfunk/mkdocs-material/blob/fdeafbb875fe7cf340da429c90e8eaa965e9663e/src/templates/partials/integrations/analytics/google.html#L23-L26
https://github.com/squidfunk/mkdocs-material/blob/fdeafbb875fe7cf340da429c90e8eaa965e9663e/src/templates/partials/integrations/analytics/google.html#L36

---

Therefore you can add:

```yml
extra:
  analytics:
    provider: custom
    authentication_id: 51laID
    special_property: 'special value'
```

And it will load the "partials/integrations/analytics/custom.html" partial, and you will be able to pass the values over to the custom script.

---

The files are Jinja2 templates that combine special syntax with HTML.
https://jinja.palletsprojects.com/en/3.0.x/templates/

**Reply by w1ndys** [reply]
I seem to understand somewhat, thank you very much for your reply!

**Comment by jamesirichard90** [comment]
What are the best ways to configure custom analytics tracking for one of our client's website: [lussoleather.com](https://www.lussoleather.com/) to gain deeper insights into user behavior and optimize our marketing efforts?

**Reply by alexvoss** [reply]
Please do not hijack an existing discussion but open a new one - with a specific question relevant to Material for MkDocs and enough information that people would actually be able to answer, as well as motivated to do so. What have you considered / tried, what do you mean by "custom" and what are the requirements and constraints?

**Reply by squidfunk** [reply]
The page doesn't even seem to be built with Material for MkDocs – I'm afraid we can't provide general web development or marketing support here for free. It might be a better idea to ask this question on StackOverflow. Thank you!
